Eval function for string_to_lower/upper_case builtin functions

romainbrenguier opened the pull request.

romainbrenguier committed:
Refactor add_axioms_for_to_lower_case

romainbrenguier committed:
Implement builtin string_to_lower_case function

romainbrenguier committed:
Refactor string_to_upper_case

romainbrenguier committed:
Add builtin class for string_to_upper_case

romainbrenguier committed:
Remove redundant function application ID check

These functions are already handled by the string dependency graph.
They should not appear in the add_axioms_for_function_application input.

romainbrenguier committed:
Tests for String.toLowerCase

romainbrenguier committed:
Tests for String.toUpperCase

peterschrammel commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -187,6 +187,25 @@ exprt string_set_char_builtin_functiont::length_constraint() const
+    if(
+      ('A' <= c && c <= 'Z') || (0xc0 <= c && c <= 0xd6) ||
+      (0xd8 <= c && c <= 0xde))
+      c += 0x20;
```
braces preferred

allredj replied:
These codes don't mean anything to me. Could you perhaps add some comments, or better, define meaningful constant names for them?

allredj replied:
OK, I see you've added some documentation. it seems I shouldn't have reviewed commit by commit. I still think defining constants would be cleaner.

peterschrammel commented on src/solvers/refinement/string_builtin_function.h:
```diff
@@ -177,6 +177,37 @@ class string_set_char_builtin_functiont
   exprt length_constraint() const override;
 };
 
+class string_to_lower_case_builtin_functiont
```
doxygen

peterschrammel commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -206,6 +206,23 @@ optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
+    return {};
+  for(mp_integer &c : input_opt.value())
+  {
+    if('a' <= c && c <= 'z')
```
why don't you consider the same set of characters as in lower_case?

romainbrenguier replied:
it's to mimic the function encoded by the constraints for these two functions. Maybe I should adapt the constraints for the to_upper_case function to work for the same characters as lower_case, otherwise that can be confusing.

peterschrammel commented on src/solvers/refinement/string_builtin_function.h:
```diff
@@ -208,6 +208,37 @@ class string_to_lower_case_builtin_functiont
   };
 };
 
+class string_to_upper_case_builtin_functiont
```
doxygen

peterschrammel commented on src/solvers/refinement/string_constraint_generator_main.cpp:
```diff
@@ -456,10 +456,6 @@ exprt string_constraint_generatort::add_axioms_for_function_application(
     res=add_axioms_for_compare_to(expr);
   else if(id==ID_cprover_string_literal_func)
     res=add_axioms_from_literal(expr);
-  else if(id==ID_cprover_string_concat_func)
```
typo in commit message: redundant

romainbrenguier replied:
fixed

peterschrammel commented on jbmc/regression/jbmc-strings/StringToLowerCase/Test.java:
```diff
@@ -0,0 +1,50 @@
+    {
+        StringBuilder builder = new StringBuilder();
+        builder.append("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".toLowerCase());
+        builder.append("abcdeghijklmnopfrstuqwxyzABCDvFGHIJKLMENOPQRSTUVWXYZ".toLowerCase());
```
do you have tests for the other letters too?

romainbrenguier replied:
No, but we don't test the result here anyway, we are just testing that unnecessary constraints are not added. I didn't find a good way to test it here: we cannot add assertions because that could change the result and recovering the string from the trace is difficult.

martin-cs replied:
A general, unhelpful point -- are we doing ASCII case conversion or Unicode case conversion, because the later, frankly, fills me with fear:

https://unicode.org/faq/casemap_charprop.html
https://www.unicode.org/reports/tr21/tr21-5.html

owen-mc-diffblue replied:
I seem to recall seeing tests for upper-casing non-ascii characters...

romainbrenguier committed:
Make to_upper_case work for Latin-1 supplement

romainbrenguier committed:
Improve documentation of add_axioms_for_to_lower_case

romainbrenguier committed:
Extract an is_upper_case function

romainbrenguier committed:
Remove assumptions that input char are < 0x100

This could conflict with other constraints and make SAT insatisfiable
leading to false positive in verification. Instead it should be to the
user of these builtin function to make sure they are used in a way that
corresponds to the function they want.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 15c01deec9bf0957a7d316a2ba29afd7db2f43c3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81007322
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 182256ac8f8e0bfb64fbd7a01db62bf80f486529).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81011045

allredj commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -175,6 +175,9 @@ class string_constraint_generatort final
     const exprt &character);
+  exprt add_axioms_for_to_lower_case(
+    const array_string_exprt &res,
+    const array_string_exprt &str);
```
Why exported?

romainbrenguier replied:
We are going to use it in the string_builtin_function object evaluation function

allredj commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -187,6 +187,25 @@ exprt string_set_char_builtin_functiont::length_constraint() const
+optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
+  const std::function<exprt(const exprt &)> &get_value) const
+{
+  auto input_opt = eval_string(input, get_value);
```
Not related to this PR, but `eval_string` could benefit from some documentation.

allredj commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -187,6 +187,25 @@ exprt string_set_char_builtin_functiont::length_constraint() const
     equal_exprt(return_code, return_value));
 }
 
+optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
```
I don't understand what the role of the `get_value` function is. Could you please document? (probably in `string_builtin_functiont`.

allredj commented on src/solvers/refinement/string_constraint_generator_transformation.cpp:
```diff
@@ -323,19 +331,14 @@ exprt string_constraint_generatort::add_axioms_for_to_lower_case(
-/// Note that index expressions are only allowed in the body of universal
-/// axioms, so we use a trivial premise and push our premise into the body.
+///   1. res.length = str.length
+///   2. forall i< str.length.
```
nit: missing space before `<`

allredj commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -178,6 +178,9 @@ class string_constraint_generatort final
     const array_string_exprt &str);
+  exprt add_axioms_for_to_upper_case(
+    const array_string_exprt &res,
+    const array_string_exprt &expr);
```
why exported?

allredj commented on src/solvers/refinement/string_builtin_function.h:
```diff
@@ -210,6 +210,8 @@ class string_to_lower_case_builtin_functiont
   };
 };
 
+/// Converting each lowercase character of Basic Latin and Latin-1 supplement
```
Nit: `Converting` -> `converting`.

allredj commented on jbmc/regression/jbmc-strings/StringToUpperCase/test_dependency.desc:
```diff
@@ -0,0 +1,10 @@
+assertion at file Test.java line 46 .*: FAILURE
+--
+--
+Check that when there are dependency, axioms are added correctly.
```
Ni: `dependency` -> `dependencies`

allredj commented on jbmc/regression/jbmc-strings/StringToUpperCase/Test.java:
```diff
@@ -0,0 +1,50 @@
+        if(s.length() < 20)
+            return "Short string";
+        if(!s.startsWith("A"))
+            return "String not starting with A";
```
I guess you planned to exclude lower case `a` here.

allredj reviewed (approved):
Looks fine to me, if a bit painful to review with all these commits that sometimes overwrite the same lines multiple times. Fixup commits can be useful for the first reviewer to check that all change request  have been implemented, but are a nuisance for the second reviewer.

martin-cs commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -187,6 +187,46 @@ exprt string_set_char_builtin_functiont::length_constraint() const
+static bool eval_is_upper_case(const mp_integer &c)
+{
+  return ('A' <= c && c <= 'Z') || (0xc0 <= c && c <= 0xd6) ||
+         (0xd8 <= c && c <= 0xde);
```
This says to me we're not doing ASCII.  Might you add a comment to say where these numbers come from?  Also, is there a reason to not use isupper et al.?

romainbrenguier replied:
Yes I documented that so that it should be clearer now. We do a bit more that ASCII but not all unicode (that dould be quite complicated indeed). My idea is that we will provide this builtin function accessible as `CProverString.toUpperCase`, which only coincide with String.toUpperCase for a fragment of unicode (basic latin and latin-1 supplement for now), you can then provide a model of String.toUpperCase which either over-approximate (useful for verification) or under-approximate (useful for test generation) the behaviour of the JDK function.

tautschnig replied:
This sounds scary. Please add tests and documentation (really, both!) that very precisely document the boundaries of what is/is not covered. I accept that there may not be enough time to do a perfect job (cover all of unicode), but then there needs to be clear documentation of what has (not) been done to avoid surprises.

martin-cs replied:
Yes; please, what @tautschnig said.

martin-cs commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -187,6 +187,46 @@ exprt string_set_char_builtin_functiont::length_constraint() const
+  for(mp_integer &c : input_opt.value())
+  {
+    if(eval_is_upper_case(c))
+      c += 0x20;
```
Does that always work?  Does it work better than to_lower?

romainbrenguier replied:
It mimics the behavior of the function specified by the constraints, which is not exactly the same as to_lower

martin-cs commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -187,6 +187,46 @@ exprt string_set_char_builtin_functiont::length_constraint() const
+    return {};
+  for(mp_integer &c : input_opt.value())
+  {
+    if(eval_is_upper_case(c - 0x20))
```
Same again.

martin-cs commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -187,6 +187,46 @@ exprt string_set_char_builtin_functiont::length_constraint() const
+         (0xd8 <= c && c <= 0xde);
+}
+
+optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
```
I'm probably not paying attention, but why not just return the exprt anyway?  This is how the rewriter works.

romainbrenguier replied:
If the eval function is not successful, the solver will get the valuation in another way (calling get of the underlying solver and doing some post-processing).

martin-cs replied:
OK... is this called before or after solving?

romainbrenguier replied:
It's called from `get` at the end of each string refinement step after solving by the underlying solver.

martin-cs reviewed (commented):
I'm unclear on what the "right" answers should be.  Is this for ASCII (in which case some parts are not needed), Unicode (in which case I am not qualified to comment but it doesn't look compliant), some Java standard (likewise), etc. ?

romainbrenguier committed:
Document eval_string

romainbrenguier committed:
Document eval_is_upper_case

romainbrenguier committed:
Document string_builtin_function::eval

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a44468bea3d1ff3a482bb3528c0cb0058253f549).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81049758
